libraries/state_history/include/eosio/state_history/log.hpp

```diff
@@ -216,12 +216,22 @@ class state_history_log {
          }
       }
 
-      if (block_num < _end_block)
-         truncate(block_num); //truncate is expected to always leave file pointer at the end
-      else if (!prune_config)
+      if (block_num < _end_block) {
+         static uint32_t start_block_num = block_num;
+         EOS_ASSERT( block_num > 2, chain::plugin_exception, "Existing ship log with ${eb} blocks when starting from genesis block ${b}",
+                     ("eb", _end_block)("b", block_num) );
+         if ( block_num < _begin_block || get_block_id_i(block_num) != header.block_id ) {
```

Comment:
Add comments 
- about why assert - start from genesis, don't want to truncate large ship logs
- `block_num < _begin_block` - this a pruned log and we need to call truncate to allow appending
- `get_block_id_i` - if we already have the block then no need to remove it from log only to re-write it to the log

Reply:
`get_block_id_i` reads the id from the log. It may not match `header.block_id` if a fork occurred. If they match then we are rewriting the same info. If they differ then a fork occurred and we need to overwrite the old data.

Reply:
Updated comments.